djangoheads/management/commands/waiter.py

```diff
@@ -0,0 +1,276 @@
+import importlib
+import os
+import random
+import re
+import sys
+import time
+from argparse import ArgumentParser
+from collections import deque
+from pathlib import Path
+from typing import Any, Callable, Deque, Dict, Optional, Tuple, cast
+
+from celery.app.control import Inspect
+from django.conf import settings
+from django.core.cache import InvalidCacheBackendError, caches
+from django.core.exceptions import ImproperlyConfigured
+from django.core.management import call_command
+from django.core.management.base import BaseCommand
+from django.db import connections
+from django.db.backends.utils import CursorWrapper  # noqa
+
+
+class Command(BaseCommand):
+    """Waiter for system checks before starting the main process.
+
+    Checks the availability of project's services dependencies.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._attempts = 6
+        self._timeout = 10
+        self._check_funcs: Deque[Callable[..., bool]] = deque()
+
+    help = (  # noqa: A003
+        "Waiter for system checks before starting the main process. "
+        "Checks the availability of project's services dependencies."
+    )
+
+    def add_arguments(self, parser: ArgumentParser) -> None:
+        """Add arguments to command."""
+        parser.add_argument(
+            "-n",
+            action="store",
+            type=int,
+            default=self._attempts,
+            required=False,
+            help=f"Attempts count (default {self._attempts})",
+        )
+        parser.add_argument(
+            "-t",
+            action="store",
+            type=int,
+            default=self._timeout,
+            help=f"Timeout in seconds between attempts (default {self._timeout})",
+            required=False,
+        )
+
+    def handle(self, *args: Tuple[Any, ...], **options: Dict[str, Any]) -> None:  # noqa: ARG002
+        """Command implementation."""
+        self._attempts = cast(int, options.get("n", self._attempts))
+        self._timeout = cast(int, options.get("t", self._timeout))
+
+        self._check_funcs.append(self.check_db_read)
+        # self._check_funcs.append(self.check_db_write) # TODO: Later we will add this check
+        self._check_funcs.append(self.check_migrations_are_applied)
+        self._check_funcs.append(self.check_caches)
+        self._check_funcs.append(self.check_celery)
+
+        while self._check_funcs and self._attempts > 0:
+            check_func = self._check_funcs.popleft()
+            res = check_func()
+            if not res:
+                self._check_funcs.appendleft(check_func)
+                self._attempts -= 1
+                if self._attempts > 0:
+                    time.sleep(self._timeout)
+                    continue
+                sys.exit(1)
+
+    def check_db_read(self) -> bool:
```

Comment:
Every test should be presented as a separate file with same interface. That will help test better and define different usage entrypoints